src/orange/orangeinp/PolySolid.cc

```diff
@@ -0,0 +1,252 @@
+
+            // Subtract (i.e., "and not") inner shape from this segment
+            auto sub_node = vb.insert_region({}, Negated{inner_node});
+            segment_node
+                = vb.insert_region(Label{label, std::to_string(i)},
+                                   Joined{op_and, {segment_node, sub_node}});
+        }
+        segment_nodes.push_back(segment_node);
+    }
+
+    // Union the given segments to create a new CSG node
+    return vb.insert_region(Label{label, "segments"},
+                            Joined{op_or, std::move(segment_nodes)});
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct an enclosed angle if applicable.
+ */
+[[nodiscard]] NodeId construct_enclosed_angle(PolySolidBase const& base,
+                                              detail::VolumeBuilder& vb,
+                                              NodeId result)
+{
+    if (auto const& sea = base.enclosed_angle())
+    {
+        // The enclosed angle is "true" (specified by the user to truncate the
+        // shape azimuthally): construct a wedge to be added or deleted
+        auto&& [sense, wedge] = sea.make_wedge();
+        NodeId wedge_id
+            = build_intersect_region(vb, base.label(), "angle", wedge);
+        if (sense == Sense::outside)
+        {
+            wedge_id = vb.insert_region({}, Negated{wedge_id});
+        }
+        result
+            = vb.insert_region(Label{std::string{base.label()}, "restricted"},
+                               Joined{op_and, {result, wedge_id}});
+    }
+    return result;
+}
+
+//---------------------------------------------------------------------------//
+}  // namespace
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct from a filled polygon solid.
+ */
+PolySegments::PolySegments(VecReal&& outer, VecReal&& z)
+    : PolySegments{{}, std::move(outer), std::move(z)}
+{
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct from a shell of a polygon solid.
+ */
+PolySegments::PolySegments(VecReal&& inner, VecReal&& outer, VecReal&& z)
+    : inner_{std::move(inner)}, outer_{std::move(outer)}, z_{std::move(z)}
+{
+    CELER_VALIDATE(z_.size() >= 2,
+                   << "no axial segments was specified: at least 2 points "
+                      "needed (given "
+                   << z_.size() << ")");
+    CELER_VALIDATE(outer_.size() == z_.size(),
+                   << "inconsistent outer radius size (" << outer_.size()
+                   << "): expected " << z_.size());
+    CELER_VALIDATE(inner_.empty() || inner_.size() == z_.size(),
+                   << "inconsistent inner radius size (" << inner_.size()
+                   << "): expected " << z_.size());
+
+    CELER_VALIDATE(is_monotonic_nondecreasing(make_span(z_)),
+                   << "axial grid has increasing grid points");
```

Comment:
```suggestion
                   << "axial grid has decreasing grid points");
```

Reply:
Nope, "is not all nondecreasing" = "has increasing"

Reply:
Yes, I know.  Still, my last two comments above hold.

Reply:
Ok I'll take a closer look 😅

Reply:
Goodness I am so obtuse. I am so sorry for doubting you, I clearly have not had enough sleep!! I'll fix it in the polyprism pull request.